src/fsharp/FSharp.Core/array.fs

```diff
@@ -494,12 +494,23 @@ namespace Microsoft.FSharp.Collections
 
         [<CompiledName("Filter")>]
         let filter f (array: _[]) = 
-            checkNonNull "array" array
-            let res = List<_>() // ResizeArray
-            for i = 0 to array.Length - 1 do 
-                let x = array.[i] 
-                if f x then res.Add(x)
-            res.ToArray()
+            checkNonNull "array" array                        
+            let temp = Array.zeroCreateUnchecked array.Length            
```

Comment:
@ReedCopsey 
Perf is fairly important to these type of apis, and indexing a bitarray is quite expensive.  It is important to reduce allocations, not necessarily allocated bytes.